[CI] InferenceWithSecurityRestIT test {p0=inference/80_region_policy_crud/Get region policy allowed with monitor_inference privilege} failing

**Build Scans:**
- [elasticsearch-intake #42575 / part4](https://gradle-enterprise.elastic.co/s/7nxyfol5kxotk)
- [elasticsearch-pull-request #159689 / part-4](https://gradle-enterprise.elastic.co/s/qhb667p62ipgy)
- [elasticsearch-pull-request #159683 / part-4](https://gradle-enterprise.elastic.co/s/tdu355ojhgvpq)
- [elasticsearch-pull-request #159692 / part-4](https://gradle-enterprise.elastic.co/s/gkqe6urdhdupy)
- [elasticsearch-pull-request #159689 / part-4](https://gradle-enterprise.elastic.co/s/7uwwxyyu47w7i)
- [elasticsearch-pull-request #159683 / part-4](https://gradle-enterprise.elastic.co/s/eoqq7wy2kdfiw)
- [elasticsearch-pull-request #159663 / part-4](https://gradle-enterprise.elastic.co/s/m2jzf7ejyyry2)
- [elasticsearch-pull-request #159641 / part-4](https://gradle-enterprise.elastic.co/s/t6hos6mafcdqg)
- [elasticsearch-pull-request #159648 / part-4](https://gradle-enterprise.elastic.co/s/i3wkghaozufkc)
- [elasticsearch-pull-request #159590 / part-4](https://gradle-enterprise.elastic.co/s/kkhgznzf5itqc)

**Reproduction Line:**
```
./gradlew ":x-pack:plugin:inference:qa:inference-with-security:yamlRestTest" --tests "org.elasticsearch.xpack.inference.InferenceWithSecurityRestIT.test {p0=inference/80_region_policy_crud/Get region policy allowed with monitor_inference privilege}" -Dtests.seed=1E631F5253F7D78F -Dtests.locale=tt -Dtests.timezone=Canada/Saskatchewan -Druntime.java=26
```

**Applicable branches:**
main

**Reproduces locally?:**
N/A

**Failure History:**
[See dashboard](https://es-delivery-stats.elastic.dev/app/dashboards#/view/dcec9e60-72ac-11ee-8f39-55975ded9e63?_g=(refreshInterval:(pause:!t,value:60000),time:(from:now-7d%2Fd,to:now))&_a=(controlGroupState:(initialChildControlState:('0c0c9cb8-ccd2-45c6-9b13-96bac4abc542':(dataViewId:fbbdc689-be23-4b3d-8057-aa402e9ed0c5,fieldName:task.keyword,order:0,selectedOptions:!(),title:'GradleTask',type:optionsListControl),'4e6ad9d6-6fdc-4fcc-bf1a-aa6ca79e0850':(dataViewId:fbbdc689-be23-4b3d-8057-aa402e9ed0c5,fieldName:className.keyword,order:1,selectedOptions:!(org.elasticsearch.xpack.inference.InferenceWithSecurityRestIT),title:'Suite',type:optionsListControl),'144933da-5c1b-4257-a969-7f43455a7901':(dataViewId:fbbdc689-be23-4b3d-8057-aa402e9ed0c5,fieldName:name.keyword,order:2,selectedOptions:!('test%20%7Bp0%3Dinference%2F80_region_policy_crud%2FGet%20region%20policy%20allowed%20with%20monitor_inference%20privilege%7D'),title:'Test',type:optionsListControl)))))

**Failure Message:**
```
java.lang.AssertionError: java.lang.AssertionError: Error executing section at [inference/80_region_policy_crud:49]: Connection is closed
```

**Issue Reasons:**
- [main] 1 consecutive failures in test test {p0=inference/80_region_policy_crud/Get region policy allowed with monitor_inference privilege}
- [main] 1 consecutive failures in step part4
- [main] 1 consecutive failures in pipeline elasticsearch-intake
- [main] 17 failures in test test {p0=inference/80_region_policy_crud/Get region policy allowed with monitor_inference privilege} (7.4% fail rate in 229 executions)
- [main] 2 failures in step part4 (5.9% fail rate in 34 executions)
- [main] 15 failures in step part-4 (12.2% fail rate in 123 executions)
- [main] 2 failures in pipeline elasticsearch-intake (5.9% fail rate in 34 executions)
- [main] 12 failures in pipeline elasticsearch-pull-request (10.7% fail rate in 112 executions)

**Note:**
This issue was created using new test triage automation. Please report issues or feedback to es-delivery.

--- a/muted-tests.yml
+++ b/muted-tests.yml
@@ -782,9 +782,6 @@ tests:
 - class: org.elasticsearch.xpack.security.SecurityRolesMultiProjectIT
   method: testUpdatingFileBasedRoleAffectsAllProjects
   issue: https://github.com/elastic/elasticsearch/issues/152582
-- class: org.elasticsearch.xpack.inference.InferenceWithSecurityRestIT
-  method: test {p0=inference/80_region_policy_crud/Get region policy allowed with monitor_inference privilege}
-  issue: https://github.com/elastic/elasticsearch/issues/152587
 - class: org.elasticsearch.xpack.esql.CsvIT
   method: test {csv-spec:inlinestats.afterWhere}
   issue: https://github.com/elastic/elasticsearch/issues/152591

--- a/x-pack/plugin/inference/src/internalClusterTest/java/org/elasticsearch/xpack/inference/integration/ModelRegistryIT.java
+++ b/x-pack/plugin/inference/src/internalClusterTest/java/org/elasticsearch/xpack/inference/integration/ModelRegistryIT.java
@@ -22,6 +22,7 @@ import org.elasticsearch.cluster.service.ClusterService;
 import org.elasticsearch.common.io.stream.StreamOutput;
 import org.elasticsearch.common.settings.Settings;
 import org.elasticsearch.core.TimeValue;
+import org.elasticsearch.features.FeatureService;
 import org.elasticsearch.index.IndexNotFoundException;
 import org.elasticsearch.index.engine.VersionConflictEngineException;
 import org.elasticsearch.index.mapper.vectors.DenseVectorFieldMapper;
@@ -1018,6 +1019,8 @@ public class ModelRegistryIT extends ESSingleNodeTestCase {
 
     private void storeModelDirectlyInIndexWithoutRegistry(Model model) {
         var listener = new PlainActionFuture<BulkResponse>();
+        var clusterState = clusterAdmin().prepareState(TEST_REQUEST_TIMEOUT).get().getState();
+        var featureService = getInstanceFromNode(FeatureService.class);
 
         client().prepareBulk()
             .setRefreshPolicy(WriteRequest.RefreshPolicy.IMMEDIATE)
@@ -1027,7 +1030,8 @@ public class ModelRegistryIT extends ESSingleNodeTestCase {
                     InferenceIndex.INDEX_NAME,
                     model.getConfigurations(),
                     false,
-                    true,
+                    clusterState,
+                    featureService,
                     client()
                 )
             )
@@ -1037,7 +1041,8 @@ public class ModelRegistryIT extends ESSingleNodeTestCase {
                     InferenceSecretsIndex.INDEX_NAME,
                     model.getSecrets(),
                     false,
-                    true,
+                    clusterState,
+                    featureService,
                     client()
                 )
             )
@@ -1258,6 +1263,8 @@ public class ModelRegistryIT extends ESSingleNodeTestCase {
 
     private void storeCorruptedModel(Model model, boolean storeSecrets) {
         var listener = new PlainActionFuture<BulkResponse>();
+        var clusterState = clusterAdmin().prepareState(TEST_REQUEST_TIMEOUT).get().getState();
+        var featureService = getInstanceFromNode(FeatureService.class);
 
         client().prepareBulk()
             .setRefreshPolicy(WriteRequest.RefreshPolicy.IMMEDIATE)
@@ -1267,7 +1274,8 @@ public class ModelRegistryIT extends ESSingleNodeTestCase {
                     storeSecrets ? InferenceSecretsIndex.INDEX_NAME : InferenceIndex.INDEX_NAME,
                     storeSecrets ? model.getSecrets() : model.getConfigurations(),
                     false,
-                    true,
+                    clusterState,
+                    featureService,
                     client()
                 )
             )

--- a/x-pack/plugin/inference/src/main/java/org/elasticsearch/xpack/inference/registry/ModelRegistry.java
+++ b/x-pack/plugin/inference/src/main/java/org/elasticsearch/xpack/inference/registry/ModelRegistry.java
@@ -635,8 +635,7 @@ public class ModelRegistry implements ClusterStateListener {
             preventDeletionLock.add(inferenceEntityId);
         }
 
-        boolean includeDocType = InferencePlugin.INFERENCE_REGION_POLICY_FEATURE_FLAG.isEnabled()
-            && InferenceIndex.inferenceIndexHasV4Mappings(clusterService.state(), featureService);
+        var clusterState = clusterService.state();
         SubscribableListener.<BulkResponse>newForked((subListener) -> {
             // in this block, we try to update the stored model configurations
             var configRequestBuilder = createIndexRequestBuilder(
@@ -644,7 +643,8 @@ public class ModelRegistry implements ClusterStateListener {
                 InferenceIndex.INDEX_NAME,
                 newModel.getConfigurations(),
                 true,
-                includeDocType,
+                clusterState,
+                featureService,
                 client
             );
 
@@ -687,7 +687,8 @@ public class ModelRegistry implements ClusterStateListener {
                     InferenceSecretsIndex.INDEX_NAME,
                     newModel.getSecrets(),
                     true,
-                    includeDocType,
+                    clusterState,
+                    featureService,
                     client
                 );
 
@@ -711,7 +712,8 @@ public class ModelRegistry implements ClusterStateListener {
                     InferenceIndex.INDEX_NAME,
                     existingModel.getConfigurations(),
                     true,
-                    includeDocType,
+                    clusterState,
+                    featureService,
                     client
                 );
 
@@ -827,8 +829,7 @@ public class ModelRegistry implements ClusterStateListener {
         }
 
         var bulkRequestBuilder = client.prepareBulk().setRefreshPolicy(WriteRequest.RefreshPolicy.IMMEDIATE);
-        boolean includeDocType = InferencePlugin.INFERENCE_REGION_POLICY_FEATURE_FLAG.isEnabled()
-            && InferenceIndex.inferenceIndexHasV4Mappings(clusterService.state(), featureService);
+        var clusterState = clusterService.state();
 
         for (var model : models) {
             bulkRequestBuilder.add(
@@ -837,7 +838,8 @@ public class ModelRegistry implements ClusterStateListener {
                     InferenceIndex.INDEX_NAME,
                     model.getConfigurations(),
                     allowOverwriting,
-                    includeDocType,
+                    clusterState,
+                    featureService,
                     client
                 )
             );
@@ -848,7 +850,8 @@ public class ModelRegistry implements ClusterStateListener {
                     InferenceSecretsIndex.INDEX_NAME,
                     model.getSecrets(),
                     allowOverwriting,
-                    includeDocType,
+                    clusterState,
+                    featureService,
                     client
                 )
             );
@@ -1247,9 +1250,13 @@ public class ModelRegistry implements ClusterStateListener {
         String indexName,
         ToXContentObject body,
         boolean allowOverwriting,
-        boolean includeDocType,
+        ClusterState clusterState,
+        FeatureService featureService,
         Client client
     ) {
+        boolean includeDocType = indexName.equals(InferenceIndex.INDEX_NAME)
+            && InferencePlugin.INFERENCE_REGION_POLICY_FEATURE_FLAG.isEnabled()
+            && InferenceIndex.inferenceIndexHasV4Mappings(clusterState, featureService);
         try (XContentBuilder xContentBuilder = XContentFactory.jsonBuilder()) {
             Map<String, String> params = includeDocType
                 ? Map.of(

